### PR TITLE
Tyler #77 + bug fixes + refactoring

### PR DESCRIPTION
**Pull request by gwwatkin**
As per offline discussion, getting the tests to pass. Done some refactoring along the way.

Supersedes #77

### CODE REVIEW

File: regression_tests/cases/distillation_settings/distillation_times.1.case.sh [hunk 1]
```diff
@@ -4,7 +4,11 @@ include "qelib1.inc";
 
 qreg q[2];
 
-t q[1];
```
Comment (gwwatkin) [inline regression_tests/cases/distillation_settings/distillation_times.1.case.sh:7]: The problem here was in the waiting mechanism. Incidentally this CR fixes it, but modifying this test not to cover the waiting feature

File: src/patches/dense_patch_computation.cpp [hunk 2]
```diff
@@ -147,15 +147,6 @@ bool merge_patches(
 }
```
Comment (gwwatkin) [inline src/patches/dense_patch_computation.cpp:147]: This file is where most of the changes are ... github doesn't even want to show it by default :joy:

Comment (gwwatkin) [reply src/patches/dense_patch_computation.cpp:147]: Mostly added debugging to exceptions and tried to make intents more clear with more detailed naming. Noting where there is a logic change

File: src/patches/dense_patch_computation.cpp [hunk 3]
```diff
@@ -182,15 +173,11 @@ InstructionApplicationResult try_apply_local_instruction(
...
-        bell_state[0].id = bellprep->side1; bell_state[1].id = bellprep->side2;
 
-        slice.place_sparse_patch(bell_state[0], false);
-        slice.place_sparse_patch(bell_state[1], false);
-        slice.get_boundary_between(bellprep->cell1,bellprep->cell2)->get().is_active=true;
-        slice.get_boundary_between(bellprep->cell2,bellprep->cell1)->get().is_active=true;
+        slice.place_sparse_patch(LayoutHelpers::basic_square_patch(bellprep->cell1, bellprep->side1), false);
+        slice.place_sparse_patch(LayoutHelpers::basic_square_patch(bellprep->cell2, bellprep->side2), false);
+        slice.get_boundary_between_or_fail(bellprep->cell1, bellprep->cell2).get().is_active=true;
+        slice.get_boundary_between_or_fail(bellprep->cell2, bellprep->cell1).get().is_active=true;
```
Comment (gwwatkin) [inline src/patches/dense_patch_computation.cpp:180]: Was undefined behaviour if the cell was not found

File: src/patches/dense_patch_computation.cpp [hunk 4]
```diff
@@ -446,8 +427,8 @@ InstructionApplicationResult try_apply_instruction_direct_followup(
                         });
 
                 if(could_not_find_space_for_patch)
-                    return {std::make_unique<std::runtime_error>(
-                            "Could not find space to place patch after BusyRegion clears"),{std::move(instruction)}};
+                    return {std::make_unique<std::runtime_error>(lstk::cat(
+                            instruction,"; Could not find space to place patch after BusyRegion clears")),{}};
```
Comment (gwwatkin) [inline src/patches/dense_patch_computation.cpp:431]: Now there is no more instance of both an error and follow up instructions. This data-structure should really be refactored into a variant. I assume you have uncommited changes though, so leaving this change for another time to reduce merge conflicts on your end.

File: src/patches/dense_patch_computation.cpp [hunk 5]
```diff
@@ -463,38 +444,20 @@ InstructionApplicationResult try_apply_instruction_direct_followup(
         {
             for(const auto& occupied_cell: busy_region->region.cells)
                 if(!slice.is_cell_free(occupied_cell.cell))
-                    return {std::make_unique<std::runtime_error>("Could not find free cell in BusyRegion"),
-                            {std::move(instruction)}};
```
Comment (gwwatkin) [inline src/patches/dense_patch_computation.cpp:465]: This move and the one below were broken by changing instruction to be a reference. Was causing one of the bugs in the dag pipeline folder

File: src/patches/dense_patch_computation.cpp [hunk 6]
```diff
@@ -529,12 +492,16 @@ void run_through_dense_slices_streamed(
...
         }
         else if (application_result.maybe_error)
         {
             slice_visitor(slice);
-            throw std::runtime_error{application_result.maybe_error->what()};
+            if (instruction.wait_at_most_for == 0)
+                throw std::runtime_error{application_result.maybe_error->what()};
+            instruction.wait_at_most_for--;
+            future_instructions.push_front(instruction);
+            advance_slice(slice, layout);
```
Comment (gwwatkin) [inline src/patches/dense_patch_computation.cpp:504]: Moved here the followup logic to completely remove the use of both error and followup instructions in the application result

**Review by tylerrleblond — APPROVED** [review]

**Pull request merged**